### PR TITLE
Automated MetaWorkflowRun creation from file sets

### PR DESCRIPTION
**Pull request by alexander-veit**
This PR adds command line functionality to create MWFRs from file sets. The following commands have been added `create-illumina-mwfr`, `create-pacbio-mwfr`, `create-hic-mwfr`, `create-ont-mwfr`, `create-fastqc-mwfr`. In order to create a MWFR for an ONT data set (file set) with accession `SMAXXX` in the production account, you would run
```
create-ont-mwfr -f SMAXXX -e staging
```
where `staging` refers to the name of the environment in your `.smaht-keys.json` file.

### CODE REVIEW

File: magma_smaht/commands/create_meta_workflow_run.py [hunk 1]
```diff
@@ -1,42 +1,107 @@
...
+@click.help_option("--help", "-h")
+@click.option(
+    "-f", "--fileset-accession", required=True, type=str, help="Fileset accession"
+)
+@click.option(
+    "-l",
+    "--length-required",
+    required=True,
+    type=int,
+    help="Requried length (can be obtained from FastQC output)",
```
Comment (B3rse) [inline magma_smaht/commands/create_meta_workflow_run.py]:
```suggestion
    help="Required length (can be obtained from FastQC output)",
```

File: magma_smaht/commands/create_meta_workflow_run.py [hunk 2]
```diff
@@ -1,42 +1,107 @@
...
+)
+@click.option(
+    "-e",
+    "--auth-env",
+    required=True,
+    type=str,
+    help="Name of environment in smaht-keys file",
+)
+def mwfr_illumina_alignment_cmd(fileset_accession, length_required, auth_env):
+    """Creates a MetaWorflowRun item in the portal for Illumina alignemnt of submitted files within a file set"""
```
Comment (B3rse) [inline magma_smaht/commands/create_meta_workflow_run.py]:
```suggestion
    """Creates a MetaWorflowRun item in the portal for Illumina alignment of submitted files within a file set"""
```

File: magma_smaht/commands/create_meta_workflow_run.py [hunk 3]
```diff
@@ -1,42 +1,107 @@
...
+)
+@click.option(
+    "-e",
+    "--auth-env",
+    required=True,
+    type=str,
+    help="Name of environment in smaht-keys file",
+)
+def mwfr_pacbio_alignment_cmd(fileset_accession, auth_env):
+    """Creates a MetaWorflowRun item in the portal for Pacbio alignment of submitted files within a file set"""
```
Comment (B3rse) [inline magma_smaht/commands/create_meta_workflow_run.py]:
```suggestion
    """Creates a MetaWorflowRun item in the portal for PacBio alignment of submitted files within a file set"""
```

File: magma_smaht/commands/create_meta_workflow_run.py [hunk 4]
```diff
@@ -1,42 +1,107 @@
...
+)
+@click.option(
+    "-e",
+    "--auth-env",
+    required=True,
+    type=str,
+    help="Name of environment in smaht-keys file",
+)
+def mwfr_hic_alignment_cmd(fileset_accession, auth_env):
+    """Creates a MetaWorflowRun item in the portal for HIC alignment of submitted files within a file set"""
```
Comment (B3rse) [inline magma_smaht/commands/create_meta_workflow_run.py]:
```suggestion
    """Creates a MetaWorflowRun item in the portal for Hi-C alignment of submitted files within a file set"""
```

File: magma_smaht/create_metawfr.py [hunk 5]
```diff
@@ -9,16 +9,206 @@
...
+    print(f"Using MetaWorkflow {mwf['accession']} ({mwf['aliases'][0]})")
+
+    file_set = get_file_set(fileset_accession, smaht_key)
+    input_arg = "input_files_bam"
+    mwfr_input = get_core_alignment_mwfr_input(file_set, input_arg, smaht_key)
+    create_and_post_mwfr(mwf["uuid"], file_set, input_arg, mwfr_input, smaht_key)
+
+
+def mwfr_hic_alignment(fileset_accession, smaht_key):
+    """Creates a MetaWorflowRun item in the portal for HIC alignment of submitted files within a file set"""
```
Comment (B3rse) [inline magma_smaht/create_metawfr.py]:
```suggestion
    """Creates a MetaWorflowRun item in the portal for Hi-C alignment of submitted files within a file set"""
```

File: magma_smaht/create_metawfr.py [hunk 6]
```diff
@@ -9,16 +9,206 @@
...
+    mwfr_input = get_core_alignment_mwfr_input_from_readpairs(
+        file_set, input_arg_1, input_arg_2, smaht_key
+    )
+    # Illumina specific input
+    mwfr_input.append(get_mwfr_parameter_input_arg("length_required", length_required))
+    create_and_post_mwfr(mwf["uuid"], file_set, input_arg_1, mwfr_input, smaht_key)
+
+
+def mwfr_pacbio_alignment(fileset_accession, smaht_key):
+    """Creates a MetaWorflowRun item in the portal for Pacbio alignment of submitted files within a file set"""
```
Comment (B3rse) [inline magma_smaht/create_metawfr.py]:
```suggestion
    """Creates a MetaWorflowRun item in the portal for PacBio alignment of submitted files within a file set"""
```

File: magma_smaht/create_metawfr.py [hunk 7]
```diff
@@ -9,16 +9,206 @@
...
+    input_arg_1 = "input_files_r1_fastq_gz"
+    input_arg_2 = "input_files_r2_fastq_gz"
+    mwfr_input = get_core_alignment_mwfr_input_from_readpairs(
+        file_set, input_arg_1, input_arg_2, smaht_key
+    )
+    create_and_post_mwfr(mwf["uuid"], file_set, input_arg_1, mwfr_input, smaht_key)
+
+
+def mwfr_ont_alignment(fileset_accession, smaht_key):
+    """Creates a MetaWorflowRun item in the portal for HIC alignment of submitted files within a file set"""
```
Comment (B3rse) [inline magma_smaht/create_metawfr.py]:
```suggestion
    """Creates a MetaWorflowRun item in the portal for ONT alignment of submitted files within a file set"""
```

File: magma_smaht/create_metawfr.py [hunk 8]
```diff
@@ -9,16 +9,206 @@
...
 
+
+def mwfr_illumina_alignment(fileset_accession, length_required, smaht_key):
+    """Creates a MetaWorflowRun item in the portal for Illumina alignemnt of submitted files within a file set"""
+
+    mwf = get_latest_mwf(MWF_NAME_ILLUMINA, smaht_key)
+    print(f"Using MetaWorkflow {mwf['accession']} ({mwf['aliases'][0]})")
+
+    file_set = get_file_set(fileset_accession, smaht_key)
+    input_arg_1 = "input_files_r1_fastq_gz"
```
Comment (B3rse) [inline magma_smaht/create_metawfr.py]: maybe we can move all these argument names we are hard coding as GLOBAL variables as well like the MWF_NAME?

File: pyproject.toml
Comment (B3rse) [inline pyproject.toml:1]: do we need to increase the version?

File: magma_smaht/utils.py [hunk 9]
```diff
@@ -134,3 +135,102 @@ def keep_last_item(items: Sequence) -> Sequence:
...
+
+def get_library_from_file_set(file_set, smaht_key):
+    """Get the library that is associated with a fileset
+
+    Args:
+        file_set(dicr): fileset from portal
+        smaht_key (dict): SMaHT key
+
+    Raises:
+        Exception: Raises an exception when there are multiple libraried associated
```
Comment (B3rse) [inline magma_smaht/utils.py]:
```suggestion
        Exception: Raises an exception when there are multiple libraries associated
```

File: magma_smaht/utils.py [hunk 10]
```diff
@@ -134,3 +135,102 @@ def keep_last_item(items: Sequence) -> Sequence:
...
+    if len(analyte["samples"]) > 1:
+        raise Exception(f"Multiple samples found for library {library['accession']}")
+    sample = ff_utils.get_metadata(
+        analyte["samples"][0], add_on="frame=raw&datastore=database", key=smaht_key
+    )
+    return sample
+
+
+def get_latest_mwf(mwf_name, smaht_key):
+    """Gets the latest version of the MWF with name `mwf_name`
```
Comment (B3rse) [inline magma_smaht/utils.py]:
```suggestion
    """Get the latest version of the MWF with name `mwf_name`
```

File: magma_smaht/utils.py [hunk 11]
```diff
@@ -134,3 +135,102 @@ def keep_last_item(items: Sequence) -> Sequence:
     elif len(items) > 1:
         result = items[-1:]
     return result
+
+
+def get_file_set(fileset_accession, smaht_key):
+    """Get the file set from its accession
```
Comment (B3rse) [inline magma_smaht/utils.py]:
```suggestion
    """Get the fileset from its accession
```

File: magma_smaht/create_metawfr.py (hunk 8 above)
Comment (alexander-veit) [reply magma_smaht/create_metawfr.py]: Done

File: pyproject.toml
Comment (alexander-veit) [reply pyproject.toml:1]: Yes

**Pull request merged**